dvc/repo/diff.py

```diff
@@ -45,6 +45,10 @@ def meta_cmp_key(meta):
         with_renames=True,
         meta_cmp_key=meta_cmp_key,
         roots=data_keys,
+        # we need to get unknown and unchanged entries to tell whether
+        # the object is missing from the cache or not.
+        with_unknown=with_missing,
+        with_unchanged=with_missing,
```

Comment:
We only check for missing cache items if the second argument is missing in `dvc diff`. So, this fix will only work in that case.